Tests for LateralProfile

MarekOchocki opened the pull request.

MarekOchocki committed:
Add tests for LateralProfile

grzanka commented:
@MarekOchocki investigate why tests are failing

MarekOchocki commented:
Tests are failing, because of issue #15

grzanka commented:
I agree with that, but what about this problem:
```
>       self.assertEqual(self.data2.penumbra_right(), 0.64)
E       AssertionError: 0.6400000000000001 != 0.64
```
?

Will you fix issue #15 in a commit belonging to this PR ?

MarekOchocki committed:
Fix tests

- change AssertEqual to AssertAlmostEqual
- change values in TestLateralNormalize

MarekOchocki committed:
Fix normalize in lateral_profile (issue #15)

MarekOchocki commented:
Yes. Last commit in this PR fixes issue #15

MarekOchocki committed:
Fix PEP

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -120,7 +120,19 @@ def normalize(self, dt, allow_cast=True):
+        slice_y = self.y[np.fabs(self.x) <= dt]
+        slice_x = self.x[np.fabs(self.x) <= dt]
+
+        if np.argwhere(self.x == -dt).size == 0:
```
Add a comment explaining why this corner case require special treatment.

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -120,7 +120,19 @@ def normalize(self, dt, allow_cast=True):
+            slice_y = np.append(self.y_at_x(-dt), slice_y)
+            slice_x = np.append(-dt, slice_x)
+
+        area = np.trapz(slice_y, slice_x)
```
Why first goes y then x ?

MarekOchocki replied:
That's normal usage of numpy.trapz() function: https://docs.scipy.org/doc/numpy/reference/generated/numpy.trapz.html

MarekOchocki committed:
Add comment explaining corner case in normalize()

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -120,7 +120,24 @@ def normalize(self, dt, allow_cast=True):
         self.x -= mid
 
-        ave = np.average(self.y[np.fabs(self.x) <= dt])
+        slice_y = self.y[np.fabs(self.x) <= dt]
```
Slice usually referers to indices, not to subset of numpy array. See i.e.:

```
>>> l = ['a', 'b', 'c', 'd' ]
>>> l[2:3]
['c']
>>> slice_tmp = slice(2,3)
>>> l[slice_tmp]
['c']
```

I'd recommend to rename this slices to sth like `norm_core_x` or `norm_section_x` (and `_y` also).

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -120,7 +120,24 @@ def normalize(self, dt, allow_cast=True):
+        That's why we interpolate points on the edge below.
+        """
+        if np.argwhere(self.x == -dt).size == 0:
+            slice_y = np.append(slice_y, self.y_at_x(dt))
```
This is expensive in terms of memory allocation. I'd avoid it.

Why not calculate first area over usuall slices.

Then do such if statements as here and add missing area, using code like:

```
if left_point_not_found:
  x_coords = (-dt, slice_x[0])
  y_coords = (self.y_at_x(-dt), slice_y[0])
  area += np.trapz(x_coords, y_coords)
```

MarekOchocki committed:
Optimize handling corner case in normalize()

MarekOchocki committed:
FiX PEP8

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -120,7 +120,26 @@ def normalize(self, dt, allow_cast=True):
+        if np.argwhere(self.x == -dt).size == 0:
+            coords_y = (self.y_at_x(-dt), norm_section_y[0])
+            coords_x = (-dt, norm_section_x[0])
+            area += np.trapz(coords_y, coords_x)
```
numpy `trapz` has a counterintuitive way of specifying parameters. I'd expect to call it like: `trapz(x,y)`, not `trapz(y,x)`.
To be more specific please use named arguments and change the order of arguments (which is possible when arguments are named):

```
area += np.trapz(x = coords_x, y = coords_y)
```

This will help future developers

MarekOchocki replied:
Done

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -120,7 +120,26 @@ def normalize(self, dt, allow_cast=True):
+        That's why we interpolate points on the edge below.
+        """
+        if np.argwhere(self.x == -dt).size == 0:
+            coords_y = (self.y_at_x(-dt), norm_section_y[0])
```
Is there a test which covers this corner case ?

MarekOchocki replied:
Yes, it is. Test coverage is currently 100%

MarekOchocki committed:
Add tests for Profiles made of integers

grzanka commented on tests/lateral.py:
```diff
@@ -35,12 +35,15 @@ def setUp(self):
         with self.assertRaises(ValueError):
             self.data3.normalize(1)
+        with self.assertRaises(TypeError):
+            self.data4.normalize(1, False)
```
Add here named arguments `dt, allow_cast`

MarekOchocki committed:
Change test for Profile of integers

MarekOchocki committed:
Add handling with Profiles made of integers

MarekOchocki committed:
Add test for normalization of normal Profile with allow_cast == False

MarekOchocki committed:
Change numpy.trapz usage to be more intuitive

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -87,17 +87,17 @@ def asymmetry(self):
             left_index_arr = np.argwhere(self.x < 0)
-            area_left += np.trapz(np.append(self.y[left_index_arr[left_index_arr.size - 1]], self.y_at_x(0)),
-                                  np.append(self.x[left_index_arr[left_index_arr.size - 1]], [.0]))
+            area_left += np.trapz(x=np.append(self.x[left_index_arr[left_index_arr.size - 1]], [.0]),
```
python supports negative indexing, last item of a table `t` can be obtained by call to `t[-1]`
It looks that `left_index_arr.size` is not needed

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -87,17 +87,17 @@ def asymmetry(self):
-            area_left += np.trapz(np.append(self.y[left_index_arr[left_index_arr.size - 1]], self.y_at_x(0)),
-                                  np.append(self.x[left_index_arr[left_index_arr.size - 1]], [.0]))
+            area_left += np.trapz(x=np.append(self.x[left_index_arr[left_index_arr.size - 1]], [.0]),
+                                  y=np.append(self.y[left_index_arr[left_index_arr.size - 1]], self.y_at_x(0)))
```
Do we really need to use `np.append` here ? Wouldn't it work with simple two-element tuple ?

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -109,18 +109,48 @@ def normalize(self, dt, allow_cast=True):
+        If division in place is not possible and allow_cast is False
+        an exception is raised.
         """
+        try:
```
Explain in comments the reason of such `try` block. Sth like:

We will try now to divide the data by 1.0. If self.y is float, then nothing happens.
In case its has integers there are two cases:
 - conversion to float will happen blabla
 - integer will stay integers bla bla

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -109,18 +109,48 @@ def normalize(self, dt, allow_cast=True):
             raise ValueError("Part of profile is missing.")
         mid = self.x_at_y(a) + w / 2.0
-        self.x -= mid
+        if allow_cast:
```
Add some explanation why sometimes in-place addition/subtraction is used and sometimes not.
Here `mid` can be float - is it the reason ? If yes, say it clearly in comments

MarekOchocki committed:
Clear code

- add comments
- remove not necessary code
